### ISSUE TITLE
[Python] Is stratified sampling performed during training?

### ISSUE THREAD

**Issue by ad1729** [Microsoft/LightGBM#1233]
Does the [`bagging_fraction` parameter](http://lightgbm.readthedocs.io/en/latest/Parameters.html#learning-control-parameters) perform stratified sampling at each iteration?

**Comment by guolinke** [comment]
it is not the stratified sampling. it is the random sampling.

**Comment by ad1729** [comment]
So for highly imbalanced datasets (1% positive class examples) and bagging_fraction = 0.8 (say), there might be iterations where there are no positive class cases in the sample of observations used at that iteration?

**Comment by guolinke** [comment]
yes, the sampling is independent with the label.

**Comment by ad1729** [comment]
Thanks for the clarification! Is this something that will be supported in the future?

**Comment by guolinke** [comment]
I am not sure about this will be useful or not. 
I think using stratified sampling will change the distribution of training data, which will cause the bad result in many cases. 
If changing the distribution of training data can help, you can just use the balanced sampled data before use GBDT, which is not need the stratified sampling as well.

**Comment by ad1729** [comment]
My point was along the following lines:

sample size: 1,000,000
positive class cases: 10,000 (so 1%)

We set `bagging fraction = 0.8` and `scale_pos_weight = 99` (and `bagging_freq = 1`)

So at each iteration M while building the ensemble, it samples 800,000 of which 8,000 are positive class cases, instead of the random sampling. Thus, the class proportions are maintained and are the same as the full sample.

> using stratified sampling will change the distribution of training data

Wouldn't this be the case when running the procedure with `bagging fraction < 1.0`?

So far I'm using `bagging fraction = 1.0`. Should I avoid using any other value for `bagging_fraction`?

**Comment by guolinke** [comment]
> it samples 800,000 of which 8,000 are positive class cases, instead of the random sampling. 

I didn't understand this. I think random sampling will produce "800,000 of which 8,000 are positive". 

"scale_pos_weight" doesn't affect the sampling.

I think `bagging fraction` is safe, it will not change the distribution.

**Comment by ad1729** [comment]
Never mind, figured it out. Thanks for the discussion!
